### ISSUE TITLE
MongoDB Prometheus exporter not scrapping database metrics

### ISSUE THREAD

**Issue by Ed87** [bitnami/containers#13533]
### Name and Version

docker.io/bitnami/mongodb:5.0.9-debian-10-r15

### What steps will reproduce the bug?

I have a Mongo deployment with a metrics exporter sidecar. When I load the associated [Percona dashboard](https://grafana.com/grafana/dashboards/16490-opstree-mongodb-dashboard/) in Grafana the database metrics are not showing - the exporter is not scrapping these metrics at all:

<img width="932" alt="MONGO-DASHBOARD-2" src="https://user-images.githubusercontent.com/23119187/201492378-52c31106-2f54-4ae1-9453-8a417117f0ca.png">

My configuration:

`apiVersion: apps/v1
kind: Deployment
metadata:
  name: mongodb-prom
  namespace: "labs"
  labels:
    app.kubernetes.io/name: mongodb
    helm.sh/chart: mongodb-12.1.12
    app.kubernetes.io/instance: mongodb-prom
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: mongodb
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxUnavailable: 1
    type: RollingUpdate  
  selector:
    matchLabels:
      app.kubernetes.io/name: mongodb
      app.kubernetes.io/instance: mongodb-prom
      app.kubernetes.io/component: mongodb
  template:
    metadata:
      labels:
        app.kubernetes.io/name: mongodb
        helm.sh/chart: mongodb-12.1.12
        app.kubernetes.io/instance: mongodb-prom
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: mongodb
    spec:
      
      serviceAccountName: mongodb-prom
      affinity:
          
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchLabels:
                    app.kubernetes.io/name: mongodb
                    app.kubernetes.io/instance: mongodb-prom
                    app.kubernetes.io/component: mongodb
                namespaces:
                  - "labs"
                topologyKey: kubernetes.io/hostname
              weight: 100
          
      securityContext:
        fsGroup: 1001
        sysctls: []
      
      containers:
        - name: mongodb
          image: docker.io/bitnami/mongodb:5.0.9-debian-10-r15
          imagePullPolicy: "IfNotPresent"
          securityContext:
            runAsNonRoot: true
            runAsUser: 1001
          env:
            - name: BITNAMI_DEBUG
              value: "false"
            - name: MONGODB_ROOT_USER
              value: "root"
            - name: MONGODB_ROOT_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: mongodb-prom
                  key: mongodb-root-password
            - name: ALLOW_EMPTY_PASSWORD
              value: "no"
            - name: MONGODB_SYSTEM_LOG_VERBOSITY
              value: "0"
            - name: MONGODB_DISABLE_SYSTEM_LOG
              value: "no"
            - name: MONGODB_DISABLE_JAVASCRIPT
              value: "no"
            - name: MONGODB_ENABLE_JOURNAL
              value: "yes"
            - name: MONGODB_PORT_NUMBER
              value: "27017"
            - name: MONGODB_ENABLE_IPV6
              value: "no"
            - name: MONGODB_ENABLE_DIRECTORY_PER_DB
              value: "no"
          ports:
            - name: mongodb
              containerPort: 27017
          
          volumeMounts:
            - name: datadir
              mountPath: /bitnami/mongodb
            - name: datadir
              mountPath: /tmp           
        - name: metrics
          image: bitnami/mongodb-exporter:latest
          imagePullPolicy: "IfNotPresent"
          securityContext:
            runAsNonRoot: true
            runAsUser: 1001
          command:
            - /bin/bash
            - -ec
          args:
            - |
              /bin/mongodb_exporter --web.listen-address ":9216" --mongodb.uri "mongodb://$MONGODB_ROOT_USER:$(echo $MONGODB_ROOT_PASSWORD | sed -r "s/@/%40/g;s/:/%3A/g")@localhost:27017/admin?" 
          env:
            - name: MONGODB_ROOT_USER
              value: "root"
            - name: MONGODB_ROOT_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: mongodb-prom
                  key: mongodb-root-password
          ports:
            - name: metrics
              containerPort: 9216
          resources:
            limits: {}
            requests: {}
      volumes:
        - name: datadir
          persistentVolumeClaim:
            claimName: mongodb`

metrics-svc.yml

`apiVersion: v1
kind: Service
metadata:
  name: mongodb-metrics
  namespace: "labs"
  labels:
    app.kubernetes.io/name: mongodb
    helm.sh/chart: mongodb-12.1.12
    app.kubernetes.io/instance: mongodb-prom
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: metrics
  annotations:
    prometheus.io/path: /metrics
    prometheus.io/port: '9216'
    prometheus.io/scrape: "true"
spec:
  type: ClusterIP
  ports:
    - port: 9216
      targetPort: metrics
      protocol: TCP
      name: http-metrics
  selector:
    app.kubernetes.io/name: mongodb
    app.kubernetes.io/instance: mongodb-prom
    app.kubernetes.io/component: mongodb`

From the dashboard side I have done an **Explore** on one of the empty metrics (i,e Received Insert Operations) and it looks like below :
<img width="530" alt="image" src="https://user-images.githubusercontent.com/23119187/201494380-5dee1f24-5d90-4e66-9371-acb11c1b53c1.png">

`"expr": "rate(mongodb_ss_opcounters{env=\"$env\", legacy_op_type=\"insert\"}[$__interval])"`

I am following the reference implementation given [here](https://blog.opstree.com/2022/08/23/a-detailed-guide-to-key-metrics-of-mongodb-monitoring/) and I have added the clusterMonitor role to the db user as below:

db.grantRolesToUser("root",[{ role: "clusterMonitor", db: "admin" }, { role: "read", db: "local" }])

NB ; When I explicitly create a mongo-exporter DB user with clusterMonitor role and do a RollingUpdate the metrics are not discoverable at all in Prometheus.

_I have updated the Prometheus and Grafana versions in the dashboard's JSON to match my installed versions (Could this affect anything?)_

My Installed Versions
prometheus:  2.34.0
grafana: 8.3.5

Default  Versions in the Dashboard (Dashboard ID 16490)
prometheus:  1.0.0
grafana: 8.5.3

### What is the expected behavior?

The metrics sidecar should scrape both the database metrics and the node/compute metrics . When I port forward the prometheus service and hit the metrics path(and port) for my endpoint I expect to see both node and DATABASE specific metrics like in the **example** below :

`# HELP mongodb_ss_uptime serverStatus.
# TYPE mongodb_ss_uptime untyped
mongodb_ss_uptime{cl_id="",cl_role="mongod",rs_state="0"} 1.042293e+06
# HELP mongodb_ss_uptimeEstimate serverStatus.
# TYPE mongodb_ss_uptimeEstimate untyped
mongodb_ss_uptimeEstimate{cl_id="",cl_role="mongod",rs_state="0"} 1.042293e+06
# HELP mongodb_ss_uptimeMillis serverStatus.
# TYPE mongodb_ss_uptimeMillis untyped
mongodb_ss_uptimeMillis{cl_id="",cl_role="mongod",rs_state="0"} 1.042293873e+09`

As long as these metrics are being scrapped I also expect the associated Percona Dashboards  to be working . For my use case I am using the dashboard on this [link](https://grafana.com/grafana/dashboards/16490-opstree-mongodb-dashboard/) and I expect to see similar visuals in my Grafana.

### What do you see instead?

1. In the Prometheus server dashboard the service endpoint is indeed discoverable in the Targets _but I see is only metrics related to compute and nothing about db statistics_. Part of the metrics that are being scrapped are as below ( reducted) :

`# HELP go_gc_cycles_automatic_gc_cycles_total Count of completed GC cycles generated by the Go runtime.
# TYPE go_gc_cycles_automatic_gc_cycles_total counter
go_gc_cycles_automatic_gc_cycles_total 1567
# HELP go_gc_cycles_forced_gc_cycles_total Count of completed GC cycles forced by the application.
# TYPE go_gc_cycles_forced_gc_cycles_total counter
go_gc_cycles_forced_gc_cycles_total 0
# TYPE go_gc_heap_allocs_by_size_bytes_total histogram
go_gc_heap_allocs_by_size_bytes_total_bucket{le="8.999999999999998"} 731868
go_gc_heap_allocs_by_size_bytes_total_bucket{le="24.999999999999996"} 6.529037e+06
go_gc_heap_allocs_by_size_bytes_total_bucket{le="64.99999999999999"} 9.475021e+06`

2. Only the Mongo UP metric is working i. e **mongodb_up{env=~""}** and looking at the IP in the dashboard panel I can deduce that  its referencing  the _metrics-svc.yml_ service and not the actual Mongo database.

### Additional information

1. I have tried using the Bitnami MongoDB Exporter version for the sidecar. It yields the exact same results :

`- name: metrics
      image: docker.io/bitnami/mongodb-exporter:0.32.0-debian-11-r5
      imagePullPolicy: "IfNotPresent"
      securityContext:
        runAsNonRoot: true
        runAsUser: 1001
      command:
        - /bin/bash
        - -ec
      args:
        - |
          /bin/mongodb_exporter --web.listen-address ":9216" --mongodb.uri "mongodb://$MONGODB_ROOT_USER:$(echo $MONGODB_ROOT_PASSWORD | sed -r "s/@/%40/g;s/:/%3A/g")@localhost:27017/admin?" 
      env:
        - name: MONGODB_ROOT_USER
          value: "root"
        - name: MONGODB_ROOT_PASSWORD
          valueFrom:
            secretKeyRef:
              name: mongodb-prom
              key: mongodb-root-password
      ports:
        - name: metrics
          containerPort: 9216
      livenessProbe:
        failureThreshold: 3
        initialDelaySeconds: 15
        periodSeconds: 5
        successThreshold: 1
        timeoutSeconds: 5
        httpGet:
          path: /
          port: metrics
      readinessProbe:
        failureThreshold: 3
        initialDelaySeconds: 5
        periodSeconds: 5
        successThreshold: 1
        timeoutSeconds: 1
        httpGet:
          path: /
          port: metrics
      resources:
        limits: {}
        requests: {}`

2. My implementation is based on the example given on this [link](https://blog.opstree.com/2022/08/23/a-detailed-guide-to-key-metrics-of-mongodb-monitoring/)

**Comment by Ed87** [comment]
Adding --collect-all and --compatible-mode worked for me with the following configuration :

`apiVersion: apps/v1
kind: Deployment
metadata:
  name: mongodb-prom
  namespace: "labs"
  labels:
    app.kubernetes.io/name: mongodb
    helm.sh/chart: mongodb-12.1.12
    app.kubernetes.io/instance: mongodb-prom
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: mongodb
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxUnavailable: 1
    type: RollingUpdate  
  selector:
    matchLabels:
      app.kubernetes.io/name: mongodb
      app.kubernetes.io/instance: mongodb-prom
      app.kubernetes.io/component: mongodb
  template:
    metadata:
      labels:
        app.kubernetes.io/name: mongodb
        helm.sh/chart: mongodb-12.1.12
        app.kubernetes.io/instance: mongodb-prom
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: mongodb
    spec:
      
      serviceAccountName: mongodb-prom
      affinity:
          
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchLabels:
                    app.kubernetes.io/name: mongodb
                    app.kubernetes.io/instance: mongodb-prom
                    app.kubernetes.io/component: mongodb
                namespaces:
                  - "labs"
                topologyKey: kubernetes.io/hostname
              weight: 100
          
      securityContext:
        fsGroup: 1001
        sysctls: []
      
      containers:
        - name: mongodb
          image: docker.io/bitnami/mongodb:5.0.9-debian-10-r15
          imagePullPolicy: "IfNotPresent"
          securityContext:
            runAsNonRoot: true
            runAsUser: 1001
          env:
            - name: BITNAMI_DEBUG
              value: "false"
            - name: MONGODB_ROOT_USER
              value: "root"
            - name: MONGODB_ROOT_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: mongodb-prom
                  key: mongodb-root-password
            - name: ALLOW_EMPTY_PASSWORD
              value: "no"
            - name: MONGODB_SYSTEM_LOG_VERBOSITY
              value: "0"
            - name: MONGODB_DISABLE_SYSTEM_LOG
              value: "no"
            - name: MONGODB_DISABLE_JAVASCRIPT
              value: "no"
            - name: MONGODB_ENABLE_JOURNAL
              value: "yes"
            - name: MONGODB_PORT_NUMBER
              value: "27017"
            - name: MONGODB_ENABLE_IPV6
              value: "no"
            - name: MONGODB_ENABLE_DIRECTORY_PER_DB
              value: "no"
          ports:
            - name: mongodb
              containerPort: 27017
          
          volumeMounts:
            - name: datadir
              mountPath: /bitnami/mongodb
            - name: datadir
              mountPath: /tmp           
        - name: metrics
          image: docker.io/bitnami/mongodb-exporter:0.32.0-debian-11-r5
          imagePullPolicy: "IfNotPresent"
          securityContext:
            runAsNonRoot: true
            runAsUser: 1001
          command:
            - /bin/bash
            - -ec
          args:
            - |
              /bin/mongodb_exporter --web.listen-address ":9216" --mongodb.uri "mongodb://$MONGODB_ROOT_USER:$(echo $MONGODB_ROOT_PASSWORD | sed -r "s/@/%40/g;s/:/%3A/g")@localhost:27017/admin?" --collect-all --compatible-mode 
          env:
            - name: MONGODB_ROOT_USER
              value: "root"
            - name: MONGODB_ROOT_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: mongodb-prom
                  key: mongodb-root-password
          ports:
            - name: metrics
              containerPort: 9216
          livenessProbe:
            failureThreshold: 3
            initialDelaySeconds: 15
            periodSeconds: 5
            successThreshold: 1
            timeoutSeconds: 5
            httpGet:
              path: /
              port: metrics
          readinessProbe:
            failureThreshold: 3
            initialDelaySeconds: 5
            periodSeconds: 5
            successThreshold: 1
            timeoutSeconds: 1
            httpGet:
              path: /
              port: metrics
          resources:
            limits: {}
            requests: {}
      volumes:
        - name: datadir
          persistentVolumeClaim:
            claimName: mongodb
       
`

**Comment by javsalgar** [comment]
Hi!

Thanks for letting us know the workaround. Are you using the Bitnami MongoDB helm chart? Is this something that should be added to the chart?

**Comment by Ed87** [comment]
@javsalgar yes im using Bitnami MongoDB helm chart (check the attached working solution). And yes , it should definitely be added to the helm chart values.yml.

**Comment by javsalgar** [comment]
Hi!

Thanks for letting us know! Would you like to submit a PR with the fix?

**Comment by Ed87** [comment]
Yes , of course. And can you please point me to the guidelines for submitting a PR.. its my first time

**Comment by javsalgar** [comment]
Thank you!! These are the Guidelines https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md

**Comment by gordes317** [comment]
@edtshuma Thank's, it's work for me!!!

**Comment by NeverBehave** [comment]
Is there actually an PR fixing this issue in bitnami Chart?
